### PR TITLE
[Feature] Distributed data collector (ray)

### PR DESCRIPTION
**Pull request by albertbou92**
## Description

new class DistributedCollector, which allows to do distributed data collection with TorchRL collectors using Ray.

@vmoens This approach required adding a new function to the collectors, one that takes a single step of the iterator. I have done it with SyncDataCollector. Do you think it is fine?  

The PR also includes an example of distributed data collection for the SyncDataCollector case.

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: torchrl/collectors/distributed/distributed_collector.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+        resources=resources)(cls)
+    c.is_remote = True
+    return c
+
+
+class DistributedCollector(IterableDataset, ABC):
+    """
+    Class to better handle the operations of ensembles of Collectors.
+
+    Contains common functionality across all collectors sets.
```
Comment (vmoens) [inline torchrl/collectors/distributed/distributed_collector.py:77]: This is not very explicit.
We should say 
"Like other collectors, this collector is an iterable that yields TensorDict instances with the appropriate number of frames."

We should say what is and what is not supported.

We should say if it's sync, async or depends on the input to the constructor.

File: torchrl/collectors/distributed/distributed_collector.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    remote_config : dict
+        Ray resource specs for the remote collectors.
+    num_collectors: int
+        Total number of collectors in the set (including local collector)
+    """
+
+    def __init__(self,
+                 collector_class,
+                 collector_params,
+                 remote_config=default_remote_config,  # TODO: mutable type as input, not correct
```
Comment (vmoens) [inline torchrl/collectors/distributed/distributed_collector.py:94]: the proper way of doing this is 
```suggestion
                 remote_config=None,
...
    if remote_config is None:
        remote_config = default_remote_config
```

File: torchrl/collectors/distributed/distributed_collector.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+                 num_collectors=1,
+                 total_frames=1000,  # TODO: is this parameter necessary? it is already specified in each collector.
+                 communication="sync",  # "sync" or "async"
+                 ):
+
+        if not _has_ray:
+            raise RuntimeError(
+                f"ray library not found, unable to create a DistributedCollector. "
+            )
```
Comment (vmoens) [inline torchrl/collectors/distributed/distributed_collector.py:103]: add `from RAY_ERR`

File: torchrl/collectors/distributed/distributed_collector.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+        num_cpus=num_cpus,
+        num_gpus=num_gpus,
+        memory=memory,
+        object_store_memory=object_store_memory,
+        resources=resources)(cls)
+    c.is_remote = True
+    return c
+
+
+class DistributedCollector(IterableDataset, ABC):
```
Comment (vmoens) [inline torchrl/collectors/distributed/distributed_collector.py:73]: Maybe `RayDistributedCollector`

File: torchrl/collectors/distributed/distributed_collector.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+        if not _has_ray:
+            raise RuntimeError(
+                f"ray library not found, unable to create a DistributedCollector. "
+            )
+
+        if communication not in ("sync", "async"):
+            raise ValueError(f"Communication parameter in CollectorSet has to be sync or async.")
+
+        # Monkey patching as_remote and print_remote_collector_info to collector class # TODO: is that ok ?
+        collector_class.as_remote = as_remote
```
Comment (vmoens) [inline torchrl/collectors/distributed/distributed_collector.py:109]: I think this is ok as long as it works :)

File: torchrl/collectors/distributed/distributed_collector.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+        self.collected_frames = 0
+        self.total_frames = total_frames
+        self.collector_class = collector_class
+        self.collector_params = collector_params
+        self.num_collectors = num_collectors
+        self.remote_config = remote_config
+        self.communication = communication
+
+        # Create a local instance of the collector class.
```
Comment (vmoens) [inline torchrl/collectors/distributed/distributed_collector.py:120]: are we sure we're not actually collecting any data? Is there a smarter way of doing this?

File: torchrl/collectors/distributed/distributed_collector.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+        return w
+
+    def add_collectors(self, num_collectors, collector_params):
+        """Create and add a number of remote collectors to the set."""
+        cls = self.collector_class.as_remote(**self.remote_config).remote
+        self._remote_collectors.extend(
+            [self._make_collector(cls, collector_params) for _ in range(num_collectors)])
+
+    def local_collector(self):
+        """Return local collector"""
```
Comment (vmoens) [inline torchrl/collectors/distributed/distributed_collector.py:147]:
```suggestion
        """Returns local collector."""
```

File: torchrl/collectors/distributed/distributed_collector.py [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+        ray.wait(pending_samples)
+
+    @staticmethod
+    def _make_collector(cls, collector_params):
+        """Create a single collector instance."""
+        w = cls(**collector_params)
+        return w
+
+    def add_collectors(self, num_collectors, collector_params):
+        """Create and add a number of remote collectors to the set."""
```
Comment (vmoens) [inline torchrl/collectors/distributed/distributed_collector.py:141]:
```suggestion
        """Creates and adds a number of remote collectors to the set."""
```

File: torchrl/collectors/distributed/distributed_collector.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+        cls = self.collector_class.as_remote(**self.remote_config).remote
+        self._remote_collectors.extend(
+            [self._make_collector(cls, collector_params) for _ in range(num_collectors)])
+
+    def local_collector(self):
+        """Return local collector"""
+        return self._local_collector
+
+    def remote_collectors(self):
+        """Returns list of remote collectors"""
```
Comment (vmoens) [inline torchrl/collectors/distributed/distributed_collector.py:151]:
```suggestion
        """Returns list of remote collectors."""
```

File: torchrl/collectors/distributed/distributed_collector.py [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+        return self._remote_collectors
+
+    def stop(self):
+        """Stop all remote collectors"""
```
Comment (vmoens) [inline torchrl/collectors/distributed/distributed_collector.py:155]:
```suggestion
        """Stops all remote collectors."""
```

File: torchrl/collectors/distributed/distributed_collector.py [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+            return self.async_iterator()
+
+    def sync_iterator(self) -> Iterator[TensorDictBase]:
+
+        while self.collected_frames < self.total_frames:
+
+            # Broadcast agent weights
+            self.local_collector().update_policy_weights_()
+            state_dict = {"policy_state_dict": self._local_collector.policy.state_dict()}
+            state_dict = ray.put(state_dict)
```
Comment (vmoens) [inline torchrl/collectors/distributed/distributed_collector.py:172]: Q: in regular data collectors we leave this to the user or ask explicitly if the update should be done at each iter.
Do we want to do the same here? ie isolating the weight updating in a dedicated method that can or won't be called at each iteration.

File: torchrl/collectors/distributed/distributed_collector.py [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+            for e in self.remote_collectors():
```
Comment (vmoens) [inline torchrl/collectors/distributed/distributed_collector.py:173]: let's use a more explicit variable name

File: torchrl/collectors/distributed/distributed_collector.py [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+                e.load_state_dict.remote(**{"state_dict": state_dict, "strict": False})
+
+            # Ask for batches to all remote workers.
+            pending_samples = [e._iterator_step.remote() for e in self.remote_collectors()]
+
+            # Wait for all rollouts
+            samples_ready = []
+            while len(samples_ready) < self.num_collectors:
+                samples_ready, samples_not_ready = ray.wait(
+                    pending_samples, num_returns=len(pending_samples), timeout=0.001)
```
Comment (vmoens) [inline torchrl/collectors/distributed/distributed_collector.py:183]: what's the role of timeout here? Is the value standard? Should it be a hyperparam or a global variable?

File: torchrl/collectors/distributed/distributed_collector.py [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+            self.collected_frames += out_td.numel()
+
+            # Update agent weights
+            self.local_collector().update_policy_weights_()
+            state_dict = {"policy_state_dict": self._local_collector.policy.state_dict()}
+            state_dict = ray.put(state_dict)
+            w.load_state_dict.remote(**{"state_dict": state_dict, "strict": False})
+
+            # Schedule a new collection task
+            future = w._iterator_step.remote()
```
Comment (vmoens) [inline torchrl/collectors/distributed/distributed_collector.py:229]: when we reach the end of the loop, we'll still have workers waiting to deliver their data right? Isn't there a risk that they will stay idle forever?

File: torchrl/collectors/distributed/distributed_collector.py (hunk 13 above)
Comment (albertbou92) [reply torchrl/collectors/distributed/distributed_collector.py:188]: If timeout is set, the function returns either when the requested number of IDs are ready or when the timeout is reached, whichever occurs first. If it is not set, the function simply waits until that number of objects is ready and returns that exact number of object refs.

Actually for not we can just set timeout=None, but if we want to implement the  preemption mechanism, we will need to check regularly how many of the tasks have ended. We can either set it to a reasonable value of allow the users to define it.

File: torchrl/collectors/distributed/distributed_collector.py (hunk 14 above)
Comment (albertbou92) [reply torchrl/collectors/distributed/distributed_collector.py:234]: Yes, I will add some code to make sure remaining tasks are cancelled or waited out.

File: torchrl/collectors/distributed/distributed_collector.py (hunk 6 above)
Comment (albertbou92) [reply torchrl/collectors/distributed/distributed_collector.py]: yes, we can just provide policy as an input parameter and assume policy will have the latest version of the weights at any moment.

File: torchrl/collectors/distributed/ray_collector.py [hunk 15]
```diff
@@ -0,0 +1,321 @@
...
+
+    def set_seed(self, seed: Union[int, List[int]], static_seed: bool = False) -> List[int]:
+        """Calls set_seed(seed, static_seed) method for each remote collector and results a list of results."""
+        if isinstance(seed, int):
+            seed = [seed]
+        if len(seed) == 1:
+            seed = seed * len(self.remote_collectors())
+        futures = [collector.set_seed.remote(s, static_seed) for s, collector in zip(seed, self.remote_collectors())]
+        results = ray.get(object_refs=futures)
+        return results
```
Comment (vmoens) [inline torchrl/collectors/distributed/ray_collector.py:298]: we should set the seed iteratively
In torchrl seeding is done via a generator that gives you the next seed to use when you need to set it multiple times
Like this you can set different seeds across workers whilst keeping the "deterministic" sequence and have seeds that don't overlap (since the alternative would be seed=0 for worker 0, seed=1 for worker 1 etc. but seeding like this will cause some overlap when you seed with seed=1 for worker 0).
See `seed_generator` for more info.
In practice you should:
- seed worker 0, get the resulting seed
- seed worker 1 with the resulting seed, get the resulting seed
- etc

File: torchrl/collectors/distributed/ray_collector.py [hunk 16]
```diff
@@ -0,0 +1,321 @@
...
+
+            # Retrieve single rollouts
+            out_td = ray.get(future)
+            ray.internal.free(future)  # should not be necessary, deleted automatically when ref count is down to 0
+            self.collected_frames += out_td.numel()
+
+            # Update agent weights
+            policy_weights_local_collector = {"policy_state_dict": self.local_policy.state_dict()}
+            policy_weights_local_collector_ref = ray.put(policy_weights_local_collector)
+            w.load_state_dict.remote(**{"state_dict": policy_weights_local_collector_ref, "strict": False})
```
Comment (vmoens) [inline torchrl/collectors/distributed/ray_collector.py:268]: we should let the user take care of that OR have a param in the constructor that specifies how often we need to update it.
Also: i am working on a param in `update_policy_weights_()` to facilitate this, something like
```
collector.update_policy_weights_(new_weights)
```
but i'm not sure it makes sense.

IMO load_state_dict should only be used to deserialize things, it's a bit inefficient.

In practice what I would see is:

E.g.
```
policy_weights = TensorDict(dict(self.policy.named_parameters()), [])
w.update_policy_weights_(policy_weights)
```

once it's implemented, oc

File: torchrl/collectors/distributed/ray_collector.py [hunk 17]
```diff
@@ -0,0 +1,321 @@
...
+
+            # Schedule a new collection task
+            future = w._iterator_step.remote()
+            pending_tasks[future] = w
+
+            yield out_td
```
Comment (vmoens) [inline torchrl/collectors/distributed/ray_collector.py:274]: this should be moved before the update of the policy weights no?
Like this first the user receives the data, then the weights are updated and things keep on running

File: torchrl/collectors/distributed/ray_collector.py [hunk 18]
```diff
@@ -0,0 +1,321 @@
...
+            # collector.__ray_terminate__.remote()  # This will kill the actor but let pending tasks finish
+            ray.kill(collector)  # This will interrupt any running tasks on the actor, causing them to fail immediately
+
+    def iterator(self):
+        if self.coordination == "sync":
+            return self._sync_iterator()
+        else:
+            return self._async_iterator()
+
+    def _sync_iterator(self) -> Iterator[TensorDictBase]:
```
Comment (vmoens) [inline torchrl/collectors/distributed/ray_collector.py:206]: where do we update the weights here?

**Review by vmoens — COMMENTED** [review]
now we need some tests 😬

File: torchrl/collectors/distributed/ray_collector.py [hunk 19]
```diff
@@ -0,0 +1,321 @@
...
+        if isinstance(state_dict, OrderedDict):
+            state_dict = [state_dict]
+        if len(state_dict) == 1:
+            state_dict = state_dict * len(self.remote_collectors())
+        for collector, state_dict in zip(self.remote_collectors(), state_dict):
+            collector.load_state_dict.remote(state_dict)
+
+    def shutdown(self):
+        """Finishes processes started by ray.init()."""
+        ray.shutdown()
```
[the remark this answers is not in the document]
Comment (albertbou92) [reply torchrl/collectors/distributed/ray_collector.py:314]: yup :)

File: torchrl/collectors/distributed/ray_collector.py (hunk 18 above)
Comment (albertbou92) [reply torchrl/collectors/distributed/ray_collector.py:202]: here:
https://github.com/pytorch/rl/blob/b3a8ec71eeeaf00998ca68ddb9216063a1dd2a1a/torchrl/collectors/distributed/ray_collector.py#L209

File: torchrl/collectors/distributed/ray_collector.py (hunk 17 above)
Comment (albertbou92) [reply torchrl/collectors/distributed/ray_collector.py:270]: makes sense

File: torchrl/collectors/collectors.py [hunk 20]
```diff
@@ -527,6 +527,33 @@ def set_seed(self, seed: int, static_seed: bool = False) -> int:
         """
         return self.env.set_seed(seed, static_seed=static_seed)
 
+    def _iterator_step(self) -> TensorDictBase:
```
[the remark this answers is not in the document]
Comment (albertbou92) [reply torchrl/collectors/collectors.py]: we can use the next function

File: torchrl/collectors/distributed/ray_collector.py (hunk 15 above)
Comment (albertbou92) [reply torchrl/collectors/distributed/ray_collector.py:294]: we can use the iterative approach

**Review by vmoens — COMMENTED** [review]
Can we align this with the other distributed collectors?

File: torchrl/collectors/distributed/ray_collector.py [hunk 21]
```diff
@@ -0,0 +1,562 @@
...
+    Returns:
+        A function that creates ray remote class instances.
+    """
+    remote_collector = ray.remote(**remote_config)(cls)
+    remote_collector.is_remote = True
+    return remote_collector
+
+
+class RayDistributedCollector(_DataCollector):
+    """Distributed data collector with Ray (https://docs.ray.io/) backend.
```
[the remark this answers is not in the document]
Comment (albertbou92) [reply torchrl/collectors/distributed/ray_collector.py]: ok!

File: torchrl/collectors/distributed/ray.py [hunk 22]
```diff
@@ -0,0 +1,637 @@
...
+    ray.remote() when called to create each remote collector actor, including collector compute
+    resources.The sum of all collector resources should be available in the cluster. Refer to Ray
+    documentation for advanced configuration of the ray.remote() method. Default kwargs are:
+
+    {
+        "num_cpus": 1,
+        "num_gpus": 0.2,
+        "memory": 2 * 1024 ** 3,
+    }
+
```
Comment (vmoens) [inline torchrl/collectors/distributed/ray.py:107]: We could do
```suggestion
    >>> kwargs = {
    ...     "num_cpus": 1,
    ...     "num_gpus": 0.2,
    ...     "memory": 2 * 1024 ** 3,
    ... }

```

**Comment by vmoens** [comment]
Sorry for the late reply

> @vmoens This approach required adding a new function to the collectors, one that takes a single step of the iterator. I have done it with SyncDataCollector. Do you think it is fine?

I can't spot it, what is it that you changed?
We have a `collector.next()` method now if that's what you need.

Also why aren't the test following the same pattern than other distributed collectors? I think it's fine if there is a good reason.

**Review by vmoens — APPROVED** [review]
LGTM, see my comment in the PR

File: examples/distributed/collectors/multi_nodes/ray_train.py
Comment (vmoens) [inline examples/distributed/collectors/multi_nodes/ray_train.py:1]: Lovely! You went the extra length with this!

**Review by vmoens — APPROVED** [review]
Awesome!
